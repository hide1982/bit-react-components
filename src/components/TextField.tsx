import React, { forwardRef, useState } from "react";
import styled from "styled-components";

interface Props {
  placeholder?: string;
  value?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  className?: string;
}

const StyledInput = styled.input`
  box-sizing: border-box;
  outline: none;
  height: 32px;
  padding: 0 12px;
  background-color: rgba(0, 0, 0, 0.2);
  border: solid 1px #333;

  :focus {
    background-color: transparent;
  }
`;

const TextField = forwardRef<HTMLInputElement, Props>((props, ref) => {
  const { className, value: initialValue = "", onChange, placeholder } = props;
  const [value, setValue] = useState(initialValue);

  return (
    <StyledInput
      className={className}
      ref={ref}
      value={value}
      placeholder={placeholder}
      onChange={(e) => {
        if (onChange) onChange(e);
        setValue(e.target.value);
      }}
    />
  );
});

export default TextField;
