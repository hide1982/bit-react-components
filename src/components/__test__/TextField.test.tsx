import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import TextField from "../TextField";
import styled from "styled-components";

describe("TextField component", () => {
  test("snapshot", () => {
    render(<TextField />);
    expect(screen.getByTestId("text-field")).toMatchSnapshot();
  });

  test("Can input text", () => {
    render(<TextField />);
    const input = screen.getByTestId("text-field");
    fireEvent.change(input, {
      target: { value: "input text" },
    });
    expect(input).toHaveValue("input text");
  });

  test("Can set placeholder", () => {
    render(<TextField placeholder="placeholder test" />);
    expect(screen.getByPlaceholderText("placeholder test")).toBeInTheDocument();
  });

  test("Can set initial value", () => {
    render(<TextField value="initial value" />);
    expect(screen.getByTestId("text-field")).toHaveValue("initial value");
  });

  it("Can override styles by styled-component", () => {
    const StyledTextField = styled(TextField)`
      background-color: #000;
    `;
    render(<StyledTextField />);
    expect(screen.getByTestId("text-field")).toHaveStyle(
      "background-color: #000"
    );
  });

  test("Can set Disabled", () => {
    render(<TextField disabled />);
    expect(screen.getByTestId("text-field")).toBeDisabled();
  });

  test("Can use ref", () => {
    const ref = React.createRef<HTMLInputElement>();
    render(<TextField ref={ref} />);
    expect(ref.current?.dataset?.testid).toBe("text-field");
  });
});
