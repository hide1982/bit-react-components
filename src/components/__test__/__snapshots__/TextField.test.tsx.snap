// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component snapshot 1`] = `
<input
  class="sc-bdfBwQ geYXpz"
  data-testid="text-field"
  value=""
/>
`;
